Windows dll symbol exports

chilarai opened the pull request:
Merge request for fixed bugs as mentioned in issue #1221

Fixes
1)  All non templated functions inside `src/include/duckdb/main` 
2) Templated functions in Appender Class
3) **ToString()** function and Templated **CreateValue** inside `src/include/duckdb/common/types/value.hpp`

These enable us to use the generated **DuckDb dll** with basic CRUD operations and read/process CSV in Memory on Windows.  I Will keep adding more as I develop more functions to test

Mytherin commented on src/include/duckdb/common/types/value.hpp:
```diff
@@ -12,14 +12,23 @@
 
+
+#ifdef _WIN32
+#define DUCKDBAPI __declspec(dllexport)
```
Can this be moved to a common header (e.g. `duckdb/common/winapi.hpp`, which can then be included from `duckdb/common/helper.hpp` and `duckdb.h`)?

Mytherin replied:
This seems to be missing an additional #ifdef that does `__declspec(dllimport)` if a certain define is not set (e.g. `DUCKDB_BUILD_LIBRARY`).  I believe this should be e.g.:


```cpp
#ifdef WIN32
#ifdef DUCKDB_BUILD_LIBRARY
#define DUCKDB_API __declspec(dllexport)
#else
#define DUCKDB_API __declspec(dllimport)
#endif
#else
#define DUCKDB_API
#endif
```

Then the `DUCKDB_BUILD_LIBRARY` define should be added to the definitions in main cmake (`src/CMakeLists.txt` -> `add_definitions(-DDUCKDB_BUILD_LIBRARY)`)

Mytherin commented on src/main/connection.cpp:
```diff
@@ -12,24 +12,26 @@
 using namespace std;
 
-Connection::Connection(DuckDB &database) : db(database), context(make_unique<ClientContext>(database)) {
+ Connection::Connection(DuckDB &database) : db(database), context(make_unique<ClientContext>(database)) {
```
Could you run the formatter (`python3 scripts/format.py master)`? It seems that the indentation here has been changed.

Mytherin commented on src/include/duckdb/main/connection_manager.hpp:
```diff
@@ -13,19 +13,28 @@
 class ConnectionManager {
 public:
-	~ConnectionManager();
+	DUCKDBAPI ~ConnectionManager();
```
The ConnectionManager is supposed to be internally used only; does this need to be exported?

Mytherin commented on src/include/duckdb/common/types/value.hpp:
```diff
@@ -12,14 +12,23 @@
 
+
+#ifdef _WIN32
+#define DUCKDBAPI __declspec(dllexport)
```
Can we rename this from `DUCKDBAPI` to `DUCKDB_API`, for consistency reasons?

chilarai replied:
We can. But I thought since it was `DUCKDBAPI` in `duckdb.h`, I thought it is what I should keep everywhere. Do you want me to rename in the `duckdb.h` file as well?

Mytherin replied:
Yes, please rename it there as well.

Mytherin commented on src/include/duckdb/common/types/value.hpp:
```diff
@@ -12,14 +12,23 @@
+
+#ifdef _WIN32
+#define DUCKDBAPI __declspec(dllexport)
+#define DUCKDBCALL __cdecl
```
`DUCKDBCALL` does not seem to be used anywhere, can we remove it?

Mytherin reviewed (commented):
Thanks for the PR, and great that you got it to work! It looks good, thanks for the effort. I have some minor comments. Besides that, I am wondering how we can test that these symbols are properly exported. Do you think we could add a test to the Github Actions on Windows that e.g. compiles an external program and links it against the created DLL?

chilarai commented:
Sure I will do the rest. I have already created a demo project with QT C++ and DuckDb (https://github.com/chilarai/qt-duckdb). Will it be fine to integrate with Duckdb Github actions or do I need to create a generic CPP project?

Mytherin commented:
Thanks! I would prefer a tiny generic C++ project without any dependencies (e.g. a single file that launches a database and fires some basic queries).

chilarai commented:
@Mytherin, I am trying to add to Github Actions for windows to test the export functions in dll, but it cant seem to locate the generated duckdb.hpp inside `src/amalgamation/duckdb.hpp`

Also tried to generate the duckdb.hpp in my local system to check the generated path (after editing the cmakelists.txt `option(AMALGAMATION_BUILD, TRUE)` ) on windows but it throws a CMake error in `add_library(duckdb SHARED "${PROJECT_SOURCE_DIR}/src/amalgamation/duckdb.cpp")` code block

Please suggest


> Thanks! I would prefer a tiny generic C++ project without any dependencies (e.g. a single file that launches a database and fires some basic queries).

Mytherin commented:
> @Mytherin, I am trying to add to Github Actions for windows to test the export functions in dll, but it cant seem to locate the generated duckdb.hpp inside src/amalgamation/duckdb.hpp

You will need to provide an include path to your compiler. Perhaps it is easier to link to theregular headers using e.g. `/Isrc/include`.

> Also tried to generate the duckdb.hpp in my local system to check the generated path (after editing the cmakelists.txt option(AMALGAMATION_BUILD, TRUE) ) on windows but it throws a CMake error in add_library(duckdb SHARED "${PROJECT_SOURCE_DIR}/src/amalgamation/duckdb.cpp") code block

Could you post the error?

chilarai commented:
I will do that in some time and update

chilarai commented:
@Mytherin I have implemented all the changes except the Github actions test. For *__declspec(dllexport)*, I have added the following in `CMakeLists.txt`
```sh
option(EXPORT_DLL_SYMBOLS "Export dll symbols on Windows, else import" TRUE)
...
if(${EXPORT_DLL_SYMBOLS})
  # For Windows DLL export symbols
  add_definitions(-DDUCKDB_BUILD_LIBRARY)
endif()
```

Also, added a few params in `.gitignore`
```
*.tlog
*.dll
*.lib
CMakeFiles/*
ub_*.cpp
```

Lastly, implemented `winapi.hpp` inside `src/include/duckdb/common` for **DUCKDB_API __declspec(dllexport)** and renamed all DUCKDBAPI to DUCKDB_API

chilarai commented:
> You will need to provide an include path to your compiler. Perhaps it is easier to link to the regular headers using e.g. `/Isrc/include`.

I will try this now. Rest, I have placed the `cppintegration.cpp` in the root of the project. Please check if that much code is sufficient for the test and also let me know if I can put the file in the root 


> Could you post the error?

1> [CMake] CMake Error at src/CMakeLists.txt:30 (add_library):
1> [CMake]   Cannot find source file:
1> [CMake] 
1> [CMake]     C:/Users/Windows/source/Repos/duckdb/src/amalgamation/duckdb.cpp
1> [CMake] 
1> [CMake]   Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
1> [CMake]   .hpp .hxx .in .txx
1> [CMake] 
1> [CMake] 
1> [CMake] CMake Error at src/CMakeLists.txt:30 (add_library):
1> [CMake]   No SOURCES given to target: duckdb
1> [CMake] 
1> [CMake] 
1> [CMake] CMake Error at src/CMakeLists.txt:31 (add_library):
1> [CMake]   No SOURCES given to target: duckdb_static
1> [CMake] 
1> [CMake] 
1> [CMake] CMake Generate step failed.  Build files cannot be regenerated correctly.

Mytherin commented:
> I will try this now. Rest, I have placed the cppintegration.cpp in the root of the project. Please check if that much code is sufficient for the test and also let me know if I can put the file in the root


I would not add the file to the root of the project, but rather to a subdirectory (e.g. `examples/embedded-c++-windows`). Otherwise the file looks fine.

> Could you post the error?

The file is not there by default, you need to run `python3 scripts/amalgamation.py` to generate it. 

Sidenote: it seems to me this PR is changing a lot of files that are not related to the PR (e.g. `test/sql/copy/parquet/data/alltypes_plain.parquet  100755 → 100644`), adding IDE specific files (`third_party/re2/DartConfiguration.tcl`) and removing files for no discernable reason ( `tools/rpkg/configure`). It also seems that something is happening that is changing all lines in files (perhaps newlines being modified). Could you fix those issues? The PR should only contain the minimal changes necessary to implement the functionality.

chilarai commented:
Sorry for that. I will pull from the master and update my fork and branch again. Then I will push the required files only.
Also yes my IDE was changing the new line characters which I will change to work with the current setup
